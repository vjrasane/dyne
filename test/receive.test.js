import Dyne from "~";

import receiver from "~/receive";

const Component = (model, props, children) => (
  <div>
    <div>{props}</div>
    <div>{model}</div>
    <div>{children}</div>
  </div>
);

describe("receive", () => {
  let receive, view;
  beforeEach(() => {
    ({ receive, view } = receiver());
  });

  it("receives single model", () => {
    const Receiver = receive("component", Component);
    const model = {
      component: "receive this"
    };
    const viewer = () => <Receiver />;
    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });

  it("receives embedded model", () => {
    const Receiver = receive(
      ["firstChild", "secondChild", "thirdChild"],
      Component
    );
    const model = {
      firstChild: {
        secondChild: {
          thirdChild: "receive this embedded"
        }
      }
    };
    const viewer = () => <Receiver />;
    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });

  it("receives multiple components", () => {
    const FirstReceiver = receive("first", Component);
    const SecondReceiver = receive("second", Component);
    const ThirdReceiver = receive("third", Component);

    const model = {
      first: "first",
      second: "second",
      third: "third"
    };

    const viewer = () => (
      <div>
        <FirstReceiver />
        <SecondReceiver />
        <ThirdReceiver />
      </div>
    );

    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });

  it("receives props", () => {
    const Receiver = receive("component", Component);
    const model = {
      component: "receive this"
    };
    const viewer = () => <Receiver prop="value" />;
    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });

  it("receives children", () => {
    const Receiver = receive("component", Component);
    const model = {
      component: "receive this"
    };
    const viewer = () => <Receiver>child</Receiver>;
    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });

  it("receives model, props and children", () => {
    const Receiver = receive("component", Component);
    const model = {
      component: "receive this"
    };
    const viewer = () => <Receiver prop="value">child</Receiver>;
    const virtualDom = view(viewer)(model);
    expect(virtualDom).toMatchSnapshot();
  });
});
