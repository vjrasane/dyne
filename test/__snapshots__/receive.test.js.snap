// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`receive receives children 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        null,
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "receive this",
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "child",
      ],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;

exports[`receive receives embedded model 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        null,
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "receive this embedded",
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;

exports[`receive receives model, props and children 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "prop": "value",
        },
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "receive this",
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "child",
      ],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;

exports[`receive receives multiple components 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            null,
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [
            "first",
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [],
          "props": Array [],
          "type": "div",
        },
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            null,
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [
            "second",
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [],
          "props": Array [],
          "type": "div",
        },
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            null,
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [
            "third",
          ],
          "props": Array [],
          "type": "div",
        },
        Object {
          "children": Array [],
          "props": Array [],
          "type": "div",
        },
      ],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;

exports[`receive receives props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "prop": "value",
        },
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "receive this",
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;

exports[`receive receives single model 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        null,
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [
        "receive this",
      ],
      "props": Array [],
      "type": "div",
    },
    Object {
      "children": Array [],
      "props": Array [],
      "type": "div",
    },
  ],
  "props": Array [],
  "type": "div",
}
`;
